File: android/app/src/main/kotlin/com/example/inet_app/vpn/EmbeddedCoreBridge.kt
package com.example.inet_app.vpn

import android.content.Context
import java.io.File
import java.util.UUID
import java.util.concurrent.ConcurrentHashMap

class MissingSingBoxBindingException(message: String) : IllegalStateException(message)

data class SingBoxLaunchOptions(
    val context: Context,
    val sessionId: String,
    val configJson: String,
    val configFile: File,
    val workingDirectory: File,
    val tunFd: Int,
    val protectSocket: (Int) -> Boolean,
)

interface AndroidSingBoxSession {
    fun stop()
}

interface AndroidSingBoxBinding {
    @Throws(Exception::class)
    fun start(options: SingBoxLaunchOptions): AndroidSingBoxSession
}

object EmbeddedCoreBridge {
    private const val DEFAULT_BINDING_CLASS = "com.example.inet_app.vpn.libbox.LibboxSingBoxBinding"
    private const val ALT_BINDING_CLASS = "com.example.inet_app.vpn.LibboxSingBoxBinding"

    @Volatile
    private var bindingOverride: AndroidSingBoxBinding? = null
    private val sessions = ConcurrentHashMap<String, AndroidSingBoxSession>()

    fun installSingBoxBinding(binding: AndroidSingBoxBinding) {
        bindingOverride = binding
    }

    fun clearSingBoxBindingOverride() {
        bindingOverride = null
    }

    @Throws(Exception::class)
    fun startSingBox(
        context: Context,
        configJson: String,
        tunFd: Int,
        protectSocket: (Int) -> Boolean,
    ): String {
        val binding = resolveSingBoxBinding(context)
        val sessionId = UUID.randomUUID().toString()
        val runtimeDir = prepareRuntimeDirectory(context, sessionId)
        val configFile = File(runtimeDir, "sing-box.json").apply { writeText(configJson) }
        val options = SingBoxLaunchOptions(
            context = context.applicationContext,
            sessionId = sessionId,
            configJson = configJson,
            configFile = configFile,
            workingDirectory = runtimeDir,
            tunFd = tunFd,
            protectSocket = protectSocket,
        )

        try {
            val session = binding.start(options)
            sessions[sessionId] = session
            return sessionId
        } catch (error: Exception) {
            runCatching { runtimeDir.deleteRecursively() }
            throw error
        }
    }

    fun stopSingBox(sessionId: String) {
        val session = sessions.remove(sessionId)
        runCatching { session?.stop() }
    }

    fun startXray(
        configJson: String,
        tunFd: Int,
        protectSocket: (Int) -> Boolean,
    ): String {
        throw UnsupportedOperationException(
            "Xray bridge is not implemented in this patch. Keep engine=sing-box until you add an Xray binding.",
        )
    }

    fun stopXray(sessionId: String) {
        // Intentionally left blank until an Xray binding exists.
    }

    private fun prepareRuntimeDirectory(context: Context, sessionId: String): File {
        val base = File(context.noBackupFilesDir, "inet-singbox")
        val runtimeDir = File(base, sessionId)
        runtimeDir.mkdirs()
        return runtimeDir
    }

    private fun resolveSingBoxBinding(context: Context): AndroidSingBoxBinding {
        bindingOverride?.let { return it }

        val candidateNames = linkedSetOf(
            context.applicationInfo.metaData?.getString("com.example.inet_app.SINGBOX_BINDING_CLASS"),
            DEFAULT_BINDING_CLASS,
            ALT_BINDING_CLASS,
        ).filterNotNull()

        for (name in candidateNames) {
            val binding = instantiateBinding(name)
            if (binding != null) {
                return binding
            }
        }

        throw MissingSingBoxBindingException(
            "No Android sing-box binding found. Add a class implementing AndroidSingBoxBinding, " +
                "for example $DEFAULT_BINDING_CLASS, link your libbox AAR/SO, and start the core there.",
        )
    }

    private fun instantiateBinding(className: String): AndroidSingBoxBinding? {
        return try {
            val clazz = Class.forName(className)
            val instance = clazz.getDeclaredConstructor().newInstance()
            instance as? AndroidSingBoxBinding
        } catch (_: ClassNotFoundException) {
            null
        } catch (error: Exception) {
            throw IllegalStateException("Failed to initialize sing-box binding $className", error)
        }
    }
}

File: ios/PacketTunnel/EmbeddedCoreBridge.swift
import Foundation
import NetworkExtension

final class IOSSingBoxSessionBox {
  let stopHandler: () -> Void

  init(stopHandler: @escaping () -> Void) {
    self.stopHandler = stopHandler
  }

  func stop() {
    stopHandler()
  }
}

struct SingBoxLaunchContext {
  let sessionId: String
  let configJSON: String
  let configURL: URL
  let workingDirectory: URL
  let packetFlow: NEPacketTunnelFlow
}

protocol IOSSingBoxBinding {
  func start(context: SingBoxLaunchContext) throws -> IOSSingBoxSessionBox
}

enum EmbeddedCoreBridgeError: LocalizedError {
  case missingSingBoxBinding
  case failedToCreateRuntimeDirectory

  var errorDescription: String? {
    switch self {
    case .missingSingBoxBinding:
      return "No iOS sing-box binding found. Add a class named LibboxSingBoxBinding that conforms to IOSSingBoxBinding and starts the embedded core."
    case .failedToCreateRuntimeDirectory:
      return "Unable to create the iOS sing-box runtime directory."
    }
  }
}

enum EmbeddedCoreBridge {
  private static var bindingOverride: IOSSingBoxBinding?
  private static var sessions: [String: IOSSingBoxSessionBox] = [:]
  private static let defaultBindingClassNames = [
    "LibboxSingBoxBinding",
    "PacketTunnel.LibboxSingBoxBinding",
    "Runner.LibboxSingBoxBinding",
  ]

  static func installSingBoxBinding(_ binding: IOSSingBoxBinding) {
    bindingOverride = binding
  }

  static func clearSingBoxBindingOverride() {
    bindingOverride = nil
  }

  static func startSingBox(configJSON: String, packetFlow: NEPacketTunnelFlow) throws -> String {
    let binding = try resolveSingBoxBinding()
    let sessionId = UUID().uuidString
    let runtimeDirectory = try prepareRuntimeDirectory(sessionId: sessionId)
    let configURL = runtimeDirectory.appendingPathComponent("sing-box.json")
    try configJSON.write(to: configURL, atomically: true, encoding: .utf8)

    do {
      let session = try binding.start(
        context: SingBoxLaunchContext(
          sessionId: sessionId,
          configJSON: configJSON,
          configURL: configURL,
          workingDirectory: runtimeDirectory,
          packetFlow: packetFlow
        )
      )
      sessions[sessionId] = session
      return sessionId
    } catch {
      try? FileManager.default.removeItem(at: runtimeDirectory)
      throw error
    }
  }

  static func stopSingBox(sessionId: String) {
    let session = sessions.removeValue(forKey: sessionId)
    session?.stop()
  }

  static func startXray(configJSON: String, packetFlow: NEPacketTunnelFlow) throws -> String {
    throw NSError(domain: "inet.vpn", code: -2, userInfo: [NSLocalizedDescriptionKey: "Xray bridge is not implemented in this patch. Keep engine=sing-box."])
  }

  static func stopXray(sessionId: String) {
    // Intentionally left blank until an Xray binding exists.
  }

  private static func prepareRuntimeDirectory(sessionId: String) throws -> URL {
    let base = FileManager.default.urls(for: .applicationSupportDirectory, in: .userDomainMask).first
      ?? FileManager.default.temporaryDirectory
    let runtime = base
      .appendingPathComponent("inet-singbox", isDirectory: true)
      .appendingPathComponent(sessionId, isDirectory: true)
    do {
      try FileManager.default.createDirectory(at: runtime, withIntermediateDirectories: true)
      return runtime
    } catch {
      throw EmbeddedCoreBridgeError.failedToCreateRuntimeDirectory
    }
  }

  private static func resolveSingBoxBinding() throws -> IOSSingBoxBinding {
    if let bindingOverride {
      return bindingOverride
    }

    for className in defaultBindingClassNames {
      if let type = NSClassFromString(className) as? NSObject.Type,
         let instance = type.init() as? IOSSingBoxBinding {
        return instance
      }
    }

    throw EmbeddedCoreBridgeError.missingSingBoxBinding
  }
}
